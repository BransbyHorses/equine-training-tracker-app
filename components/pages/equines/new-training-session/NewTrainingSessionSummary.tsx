import React, { useState, useEffect } from "react";
import dynamic from "next/dynamic";
import {
	NewSkillTrainingSessionType,
	useNewSkillTrainingSession,
} from "../../../../utils/reducers/trainingSessionReducer";
import PageTitle from "../../../PageTitle";
import { Box, Typography } from "@mui/material";
import { convertDateToString } from "../../../../utils/helpers";
import ResponsiveButton from "../../../ResponsiveButton";
import axios from "axios";

const NewTrainingSessionSuccessModal = dynamic(
	() => import("../new-training-session/NewTrainingSessionSuccessModal")
);

export const convertDayJsDateToString = (date: any) => {
	const year = date.$y;
	const month = date.$M + 1;
	const day = `${date.$D.toString().length === 1 ? "0" : ""}${date.$D}`;
	const hour = date.$H;
	const minutes = date.$m;
	const seconds = `${date.$s.toString().length === 1 ? "0" : ""}${date.$s}`;

	return `${year}-${month}-${day} ${hour}:${minutes}:${seconds}`;
};

const NewTrainingSessionSummary = () => {
	const {
		state: { newTrainingSession, trainingProgramme },
		dispatch,
	} = useNewSkillTrainingSession();
	const [submitState, setSubmitState] = useState({
		disabled: false,
		submitting: false,
		success: false,
		error: false,
	});

	useEffect(() => {
		for (const [k, v] of Object.entries(newTrainingSession)) {
			if (k === "notes") continue;
			if (!v) {
				setSubmitState({ ...submitState, disabled: true });
				break;
			}
		}
	}, []);

	const submitNewTrainingSession = async () => {
		setSubmitState({ ...submitState, submitting: true });
		axios
			.post(
				`${process.env.NEXT_PUBLIC_URL}data/training-programmes/${
					trainingProgramme!.id
				}/skill-training-session`,
				{
					...newTrainingSession,
					date: convertDayJsDateToString(newTrainingSession.date),
				}
			)
			.then(({ data }) => {
				setSubmitState({ ...submitState, submitting: false, success: true });
			})
			.catch((err) => {
				console.error(err);
				setSubmitState({ ...submitState, submitting: false, error: true });
			});
	};

	const SummaryRow = ({ title, value, notRequired, goTo }: any) => {
		return (
			<Box
				sx={{
					py: 2,
					px: 1,
					borderBottom: "1px solid lightGray",
					display: "flex",
					justifyContent: "space-between",
					alignItems: "center",
				}}
			>
				<Box>
					<Typography fontWeight={700}>{title}</Typography>
					<Typography fontWeight={300}>{value}</Typography>
				</Box>
				<Typography
					color={!value && !notRequired ? "red" : "#42a5f5"}
					sx={{ cursor: "pointer" }}
					onClick={() =>
						dispatch({
							type: NewSkillTrainingSessionType.GO_TO,
							payload: goTo,
						})
					}
				>
					<small>{!value ? "Add" : "Change"}</small>
				</Typography>
			</Box>
		);
	};

	return (
		<>
			<NewTrainingSessionSuccessModal
				isSubmitting={submitState.submitting}
				success={submitState.success}
			/>
			<PageTitle title="Check details" />
			<Box sx={{ mb: 1, pl: 2, py: 0.5, borderLeft: "5px solid lightGray" }}>
				<Typography color="gray" fontWeight={600}>
					{trainingProgramme?.equine.name}
				</Typography>
				<Typography color="gray">
					{trainingProgramme?.trainingCategory.name}
				</Typography>
			</Box>
			<Box sx={{ mb: 2, maxWidth: "500px" }}>
				<SummaryRow
					title="Date"
					value={convertDateToString(newTrainingSession.date?.toISOString())}
					goTo="date"
				/>
				<SummaryRow
					title="Skill"
					value={newTrainingSession.skill?.name}
					goTo="skillMethod"
				/>
				<SummaryRow
					title="Training Method"
					value={newTrainingSession.trainingMethod?.name}
					goTo="skillMethod"
				/>
				<SummaryRow
					title="Environment"
					value={newTrainingSession.environment?.name}
					goTo="environment"
				/>
				<SummaryRow
					title="Skill Level"
					value={newTrainingSession.progressCode}
					goTo="progress"
				/>
				<SummaryRow
					title="Comments"
					value={newTrainingSession.notes}
					notRequired
				/>
				<ResponsiveButton
					// disabled={submitState.disabled}
					onClick={submitNewTrainingSession}
					desktopstyles={{ width: "100%", mt: 3 }}
				>
					Submit
				</ResponsiveButton>
			</Box>
		</>
	);
};

export default NewTrainingSessionSummary;
