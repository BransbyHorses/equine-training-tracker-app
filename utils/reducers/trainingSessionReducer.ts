import React, { createContext } from "react";
import {
	Skill,
	SkillTrainingSession,
	TrainingEnvironment,
	TrainingMethod,
} from "../types";

enum NewSkillTrainingSessionType {
	NEXT,
	BACK,
	SET_DATE,
	SET_SKILL,
	SET_ENVIRONMENT,
	SET_PROGRESS_CODE,
	SET_TRAINING_METHOD,
	SET_TRAINING_TIME,
	SET_NOTES,
}

interface NewSkillTrainingSessionAction {
	type: NewSkillTrainingSessionType;
	payload: any;
}

const formStages = ["date", "skill_method", "progress", "summary"];

interface NewTrainingSession {
	date: string;
	skill?: Skill;
	environment?: TrainingEnvironment;
	progressCode: string;
	trainingMethod?: TrainingMethod;
	trainingTime?: string;
	notes: "";
}

const newTrainingSession: NewTrainingSession = {
	date: "",
	skill: undefined,
	environment: undefined,
	progressCode: "",
	trainingMethod: undefined,
	trainingTime: undefined,
	notes: "",
};

export const newTrainingSessionState: {
	formStage: string;
	newTrainingSession: NewTrainingSession;
} = {
	formStage: formStages[0],
	newTrainingSession,
};

export function skillTrainingSessionReducer(
	state: { formStage: string; newTrainingSession: NewTrainingSession },
	action: NewSkillTrainingSessionAction
) {
	switch (action.type) {
		case NewSkillTrainingSessionType.NEXT:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) + 1],
			};
		case NewSkillTrainingSessionType.BACK:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) - 1],
			};
		case NewSkillTrainingSessionType.SET_DATE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					date: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_SKILL:
			return { ...state, skill: action.payload };
		case NewSkillTrainingSessionType.SET_ENVIRONMENT:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					environment: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_PROGRESS_CODE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					progressCode: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_METHOD:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingMethod: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_TIME:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingTime: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_NOTES:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					notes: action.payload,
				},
			};
		default:
			return state;
	}
}
