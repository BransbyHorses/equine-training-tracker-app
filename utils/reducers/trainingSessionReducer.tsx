import axios from "axios";
import { Dayjs } from "dayjs";
import React, { Dispatch, useReducer } from "react";
import { Skill, TrainingEnvironment, TrainingMethod } from "../types";

interface NewTrainingSession {
	date?: Dayjs;
	skill?: Skill;
	environment?: TrainingEnvironment;
	progressCode: string;
	trainingMethod?: TrainingMethod;
	trainingTime?: string;
	notes: "";
}

export enum NewSkillTrainingSessionType {
	NEXT,
	BACK,
	GO_TO,
	SET_DATE,
	SET_SKILL,
	SET_ENVIRONMENT,
	SET_PROGRESS_CODE,
	SET_TRAINING_METHOD,
	SET_TRAINING_TIME,
	SET_NOTES,
	SAVE_NEW_TRAINING_SESSION,
}
interface NewSkillTrainingSessionAction {
	type: NewSkillTrainingSessionType;
	payload?: any;
}

export function skillTrainingSessionReducer(
	state: { formStage: string; newTrainingSession: NewTrainingSession },
	action: NewSkillTrainingSessionAction
) {
	switch (action.type) {
		case NewSkillTrainingSessionType.NEXT:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) + 1],
			};
		case NewSkillTrainingSessionType.BACK:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) - 1],
			};
		case NewSkillTrainingSessionType.GO_TO:
			return {
				...state,
				formStage: action.payload,
			};
		case NewSkillTrainingSessionType.SET_DATE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					date: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_SKILL:
			return { ...state, skill: action.payload };
		case NewSkillTrainingSessionType.SET_ENVIRONMENT:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					environment: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_PROGRESS_CODE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					progressCode: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_METHOD:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingMethod: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_TIME:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingTime: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_NOTES:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					notes: action.payload,
				},
			};
		default:
			return state;
	}
}

const formStages = [
	"date",
	"skillMethod",
	"environment",
	"progress",
	"summary",
	"success",
];
const newTrainingSession: NewTrainingSession = {
	date: undefined,
	skill: undefined,
	environment: undefined,
	progressCode: "",
	trainingMethod: undefined,
	trainingTime: undefined,
	notes: "",
};

interface NewTrainingSessionState {
	formStage: string;
	newTrainingSession: NewTrainingSession;
}

export const newTrainingSessionInitialState: NewTrainingSessionState = {
	formStage: formStages[0],
	newTrainingSession,
};

interface NewTrainingSessionContextProps {
	state: NewTrainingSessionState;
	dispatch: Dispatch<NewSkillTrainingSessionAction>;
}

export const NewTrainingSessionContext = React.createContext({
	state: newTrainingSessionInitialState,
	dispatch: () => null,
} as NewTrainingSessionContextProps);

export const NewTrainingSessionProvider = (props: any) => {
	const [state, dispatch] = useReducer(
		skillTrainingSessionReducer,
		newTrainingSessionInitialState
	);
	return (
		<NewTrainingSessionContext.Provider value={{ state, dispatch }}>
			{props.children}
		</NewTrainingSessionContext.Provider>
	);
};

export function useNewSkillTrainingSession() {
	const context = React.useContext(NewTrainingSessionContext);
	if (context === undefined) {
		throw new Error(
			"useNewSkillTrainingSession must be used within a NewTrainingSessionProvider"
		);
	}
	return context;
}
