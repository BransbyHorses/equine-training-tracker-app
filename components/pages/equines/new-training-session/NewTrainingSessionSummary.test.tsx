import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import NewTrainingSessionSummary, {
	convertDayJsDateToString,
} from "./NewTrainingSessionSummary";
import dayjs from "dayjs";
import { useNewSkillTrainingSession } from "../../../../utils/reducers/trainingSessionReducer";
import { render, screen } from "@testing-library/react";

jest.mock("../../../../utils/reducers/trainingSessionReducer.tsx", () => ({
	useNewSkillTrainingSession: jest.fn(),
}));

test("will convert day js date to string for request", () => {
	const dateString = "2022-10-10T16:30:11";
	const dayJsDate = dayjs(dateString);
	const formattedDate = convertDayJsDateToString(dayJsDate);
	expect(formattedDate).toBe("2022-10-10 16:30:11");
});

test("will add zero to single digit day value", () => {
	const dateString = "2022-10-09T16:30:11";
	const dayJsDate = dayjs(dateString);
	const formattedDate = convertDayJsDateToString(dayJsDate);
	expect(formattedDate).toBe("2022-10-09 16:30:11");
});

test("will add zero to single digit second value", () => {
	const dateString = "2022-10-10T16:30:09";
	const dayJsDate = dayjs(dateString);
	const formattedDate = convertDayJsDateToString(dayJsDate);
	expect(formattedDate).toBe("2022-10-10 16:30:09");
});

describe("NewTrainingSessionSummary", () => {
	it("renders with no newTrainingSession", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage: "summary",
					newTrainingSession: {
						date: undefined,
						skill: undefined,
						environment: undefined,
						progressCode: "",
						trainingMethod: undefined,
						trainingTime: 0,
						notes: "",
					},
					trainingProgramme: {
						trainingCategory: { id: 1, name: "Test Category" },
						equine: { id: 1, name: "Test Equine" },
					},
				},
			};
		});
		const tree = renderer.create(<NewTrainingSessionSummary />);
		expect(tree).toMatchSnapshot();
	});
	it("disables submit button when newTrainingSession has falsy values", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage: "summary",
					newTrainingSession: {
						date: undefined,
						skill: undefined,
						environment: undefined,
						progressCode: "",
						trainingMethod: undefined,
						trainingTime: 0,
						notes: "",
					},
					trainingProgramme: {
						trainingCategory: { id: 1, name: "Test Category" },
						equine: { id: 1, name: "Test Equine" },
					},
				},
			};
		});
		render(<NewTrainingSessionSummary />);
		expect(screen.getByRole("button", { name: "Submit" })).toBeDisabled();
	});
	it("allows submit button to be clicked when newTrainingSession values are complete", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage: "summary",
					newTrainingSession: {
						date: dayjs(),
						skill: { id: 1, name: "Test Skill" },
						environment: { id: 1, name: "Test Environment" },
						progressCode: "Ok",
						trainingMethod: { id: 1, name: "Test Training Method" },
						trainingTime: 0,
						notes: "",
					},
					trainingProgramme: {
						trainingCategory: { id: 1, name: "Test Category" },
						equine: { id: 1, name: "Test Equine" },
					},
				},
			};
		});
		render(<NewTrainingSessionSummary />);
		expect(screen.getByRole("button", { name: "Submit" })).toHaveAttribute(
			"disabled",
			""
		);
	});
});
