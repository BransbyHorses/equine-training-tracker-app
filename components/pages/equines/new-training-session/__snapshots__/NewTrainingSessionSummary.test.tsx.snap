// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTrainingSessionSummary renders with no newTrainingSession 1`] = `
[
  <div
    className="MuiBox-root css-17tcgb"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4 css-1lni2c2-MuiTypography-root"
    >
      Check details
    </h4>
  </div>,
  <div
    className="MuiBox-root css-1g8r6et"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-440ypw-MuiTypography-root"
    >
      Test Equine
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1rfn1w6-MuiTypography-root"
    >
      Test Category
    </p>
  </div>,
  <div
    className="MuiBox-root css-6mbq14"
  >
    <div
      className="MuiBox-root css-t1q99u"
    >
      <div
        className="MuiBox-root css-0"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          Date
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-b9dwom-MuiTypography-root"
        />
      </div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-1dcmq85-MuiTypography-root"
        onClick={[Function]}
      >
        <small>
          Add
        </small>
      </p>
    </div>
    <div
      className="MuiBox-root css-t1q99u"
    >
      <div
        className="MuiBox-root css-0"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          Skill
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-b9dwom-MuiTypography-root"
        />
      </div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-1dcmq85-MuiTypography-root"
        onClick={[Function]}
      >
        <small>
          Add
        </small>
      </p>
    </div>
    <div
      className="MuiBox-root css-t1q99u"
    >
      <div
        className="MuiBox-root css-0"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          Training Method
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-b9dwom-MuiTypography-root"
        />
      </div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-1dcmq85-MuiTypography-root"
        onClick={[Function]}
      >
        <small>
          Add
        </small>
      </p>
    </div>
    <div
      className="MuiBox-root css-t1q99u"
    >
      <div
        className="MuiBox-root css-0"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          Environment
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-b9dwom-MuiTypography-root"
        />
      </div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-1dcmq85-MuiTypography-root"
        onClick={[Function]}
      >
        <small>
          Add
        </small>
      </p>
    </div>
    <div
      className="MuiBox-root css-t1q99u"
    >
      <div
        className="MuiBox-root css-0"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          Skill Level
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-b9dwom-MuiTypography-root"
        />
      </div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-1dcmq85-MuiTypography-root"
        onClick={[Function]}
      >
        <small>
          Add
        </small>
      </p>
    </div>
    <div
      className="MuiBox-root css-t1q99u"
    >
      <div
        className="MuiBox-root css-0"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          Comments
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1 css-b9dwom-MuiTypography-root"
        />
      </div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-1kojf41-MuiTypography-root"
        onClick={[Function]}
      >
        <small>
          Add
        </small>
      </p>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-ttddr6-MuiButtonBase-root-MuiButton-root"
      desktopstyles={
        {
          "mt": 3,
          "width": "100%",
        }
      }
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Submit
    </button>
  </div>,
]
`;
